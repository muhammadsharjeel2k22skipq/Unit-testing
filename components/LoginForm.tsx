'use client';
import React,{ useState } from 'react';
import { SubmitHandler,useForm,FieldValues } from 'react-hook-form';
import useFormPersist from 'react-hook-form-persist';
import axios from 'axios';


const LoginForm = ({ onLogin }: { onLogin: (data: any) => void }) => {

    const [data, setData] = useState({
        email: "",
        password: "",
        firstName: "",
        lastName: ""
    });

    const { register,handleSubmit,watch,setValue } = useForm({
        defaultValues: data
    });

    useFormPersist('FromDataPresist',{
        storage: window.sessionStorage,
        watch,
        setValue
    });
    
    const formHandler:SubmitHandler<FieldValues> = async (data) => {
        // try {
        //     const res = await axios.post('http://localhost:3333/auth/signin',data);
        //     const { access_token } = res.data;
        //     const config = {
        //         headers: { Authorization: `Bearer ${access_token}` }
        //     };
        //     const user = await axios.get('http://localhost:3333/users/me',config);
        //     console.log(user.data);
        // }
        // catch (error) {
        //     console.log("FormHandler_Error",error);
        // }
        await onLogin(data);
      try {
        const res = await axios.post('/api/login', data);
        console.log(res);
        return res.data;
      }
      catch(error) {
        console.log(error);
      }

    };
 
  return (
    <div className='p-10 flex items-center justify-center flex-col gap-2'>
        <h2 className='text-xl font-semibold tracking-tight text-gray-700'>Login Form</h2>
        <form className='flex flex-col gap-4 mt-4 items-center justify-center' onSubmit={handleSubmit(formHandler)}>
            <input type='email' id='email' placeholder='Email...' {...register('email',{ required: true })} 
              className='p-2 w-[300px] outline-none border rounded-xl'
            />
            <input type='password' id='password' placeholder='Password...' {...register('password',{ required: true })} 
              className='p-2 w-[300px] outline-none border rounded-xl'
            />
            <input type='text' id='firstName' placeholder='FirstName...' {...register('firstName',{ required: true })} 
              className='p-2 w-[300px] outline-none border rounded-xl'
            />
            <input type='text' id='lastName' placeholder='LastName...' {...register('lastName',{ required: true })} 
              className='p-2 w-[300px] outline-none border rounded-xl'
            />
            <button type='submit' className='text-xl font-semibold p-2 border rounded-xl text-gray-700 w-[100px]' > 
                Sing In
            </button>
        </form>

    </div>
  )
}

export default LoginForm;
