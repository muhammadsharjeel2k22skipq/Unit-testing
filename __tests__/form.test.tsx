import '@testing-library/jest-dom';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import LoginForm  from '@/components/LoginForm';

import { setupServer } from 'msw/node';
import { rest } from 'msw';

const handlers = [
    rest.post('/api/login', async (req,res,ctx) => {
        const body = await req.json();
        return res(
            ctx.json(body),
        );
    }),
];
const server = setupServer(...handlers);


describe('LoginForm', () => {
    beforeAll(() => server.listen());
    afterEach(() => server.restoreHandlers());
    afterAll(() => server.close());

    const mockLogin = jest.fn();

  it('renders a login form correctly', async () => {
    render(<LoginForm onLogin={mockLogin} />)

    const email = screen.getByPlaceholderText(/email\.\.\./i);
    const pass = screen.getByPlaceholderText(/password\.\.\./i);
    const firstName = screen.getByPlaceholderText(/firstname\.\.\./i);
    const lastName = screen.getByPlaceholderText(/lastname\.\.\./i);

    fireEvent.change(email, { target: { value: 'farazahmed819@gmail.com' } });
    fireEvent.change(pass, { target: { value: 'fraz_928@@321' } });
    fireEvent.change(firstName, { target: { value: 'Fraz' } });
    fireEvent.change(lastName, { target: { value: 'Ahmed' } });

    const signIn = screen.getByRole('button', {
        name: /sing in/i
    });
    fireEvent.click(signIn);

    await waitFor(() => expect(mockLogin).toHaveBeenCalledTimes(1));
    await waitFor(() => expect(mockLogin).toHaveBeenCalledWith({ 
        email: "farazahmed819@gmail.com",
        firstName: "Fraz",
        lastName: "Ahmed",
        password: "fraz_928@@321",
    }));

  })
})

// console.log(document.body.innerHTML);
