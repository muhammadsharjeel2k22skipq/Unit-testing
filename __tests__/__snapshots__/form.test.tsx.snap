// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders a login form correctly 1`] = `
<input
  class="p-2 w-[300px] outline-none border rounded-xl"
  id="email"
  name="email"
  placeholder="Email..."
  type="email"
/>
`;
